atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java

```diff
@@ -45,182 +51,197 @@
+            Stopwatch stopwatch) {
+        logger.log("Call to KVS.{} on table {} took {} ms.",
+                method, tableRef, stopwatch.elapsed(TimeUnit.MILLISECONDS));
     }
 
-    private static void logTimeAndTableCount(String method, int tableCount, Stopwatch stopwatch) {
-        log.trace("Call to KVS.{} for {} tables took {} ms.",
+    private static void logTimeAndTableCount(LoggingFunction logger, String method, int tableCount,
+            Stopwatch stopwatch) {
+        logger.log("Call to KVS.{} for {} tables took {} ms.",
                 method, tableCount, stopwatch.elapsed(TimeUnit.MILLISECONDS));
     }
 
-    private static void logTimeAndTableRange(String method, String tableName, RangeRequest range, Stopwatch stopwatch) {
-        log.trace("Call to KVS.{} on table {} with range {} took {} ms.",
-                method, tableName, range, stopwatch.elapsed(TimeUnit.MILLISECONDS));
+    private static void logTimeAndTableRange(LoggingFunction logger, String method, TableReference tableRef,
+            RangeRequest range, Stopwatch stopwatch) {
+        logger.log("Call to KVS.{} on table {} with range {} took {} ms.",
+                method, tableRef, range, stopwatch.elapsed(TimeUnit.MILLISECONDS));
     }
 
-    private final KeyValueService delegate;
+    private void maybeLog(Runnable runnable, BiConsumer<LoggingFunction, Stopwatch> logger) {
+        maybeLog(() -> {
+            runnable.run();
+            return (Void) null;
+        }, logger);
+    }
 
-    private ProfilingKeyValueService(KeyValueService delegate) {
-        this.delegate = delegate;
+    private <T> T maybeLog(Supplier<T> supplier, BiConsumer<LoggingFunction, Stopwatch> logger) {
+        return maybeLog(supplier, logger, (loggingFunction, result) -> {
+        });
     }
 
-    @Override
-    public void addGarbageCollectionSentinelValues(TableReference tableRef, Iterable<Cell> cells) {
-        if (log.isTraceEnabled()) {
+    private <T> T maybeLog(Supplier<T> supplier, BiConsumer<LoggingFunction, Stopwatch> logger,
+            BiConsumer<LoggingFunction, T> additonalLoggerWithAccessToResult) {
+        if (log.isTraceEnabled() || slowlogger.isWarnEnabled()) {
             Stopwatch stopwatch = Stopwatch.createStarted();
-            delegate.addGarbageCollectionSentinelValues(tableRef, cells);
-            log.trace("Call to KVS.addGarbageCollectionSentinelValues on table {} over {} cells took {} ms.",
-                    tableRef, Iterables.size(cells), stopwatch.elapsed(TimeUnit.MILLISECONDS));
+            Optional<T> result = Optional.empty();
+            Optional<Exception> exception = Optional.empty();
+            try {
+                T res = supplier.get();
+                result = Optional.ofNullable(res);
+                return res;
+            } catch (Exception ex) {
+                exception = Optional.of(ex);
+                throw ex;
+            } finally {
```

Comment:
I'm struggling to read this finally block so I've pushed up a change that pulls it out into an inner class.